File: persistence.py
from datetime import datetime

from sqlalchemy import create_engine, Column, String, Boolean, DateTime, func, or_
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker

from tool_logging import logger

dsn = 'sqlite:///data/readform.db?check_same_thread=False'
engine = create_engine(dsn)  # , echo=True
Base = declarative_base()


class Article(Base):
    __tablename__ = "article"
    url = Column(String(2048), primary_key=True)
    agent = Column(String(128))
    saved_to_readwise = Column(Boolean, default=False)
    save_time = Column(String(128))
    readwise_resp = Column(String(1024))
    content = Column(String())

    create_time = Column(DateTime, server_default=func.now())
    update_time = Column(DateTime, server_default=func.now(), onupdate=func.now())


Session = sessionmaker(bind=engine)


def ensure_db_schema():
    from alembic.config import Config
    from alembic import command

    def run_migrations(script_location: str, dsn: str) -> None:
        logger.info('Running DB migrations in %r on %r', script_location, dsn)
        alembic_cfg = Config()
        alembic_cfg.set_main_option('script_location', script_location)
        alembic_cfg.set_main_option('sqlalchemy.url', dsn)
        command.upgrade(alembic_cfg, 'head')

    run_migrations('alembic', dsn)


def filter_old_urls(url_list: list[str]) -> list[str]:
    """Return URLs not saved to article table."""
    saved_articles = find_article(url_list)
    unsaved_articles = [item for item in url_list if item not in [item.url for item in saved_articles]]
    return unsaved_articles


def filter_saved_urls(url_list: list[str]) -> list[str]:
    """Return URLs not saved to Readwise."""
    saved_articles = find_article(url_list, only_saved=True)
    unsaved_articles = [item for item in url_list if item not in [item.url for item in saved_articles]]
    return unsaved_articles


def _now_time_str() -> str:
    return datetime.now().strftime('%Y-%m-%d %H:%M:%S')


def add_article(url: str, agent: str, content: str):
    """Add an article to article table."""
    session = Session()
    articles = find_article([url], session=session)
    if len(articles) > 0:
        # exist, update
        articles[0].agent = agent
        articles[0].content = content
        session.commit()
    else:
        # create
        article = Article(url=url, agent=agent, content=content)
        session.add(article)
        session.commit()
    pass


def mark_url_as_saved(url: str, agent: str, resp: str):
    """Mark a URL as saved to Readwise."""
    session = Session()
    articles = find_article([url], session=session)
    if len(articles) > 0:
        # exist, update
        articles[0].saved_to_readwise = True
        articles[0].save_time = _now_time_str()
        articles[0].agent = agent
        articles[0].readwise_resp = resp
        session.commit()
    else:
        # create
        article = Article(url=url, agent=agent, saved_to_readwise=True, save_time=_now_time_str(), readwise_resp=resp)
        session.add(article)
        session.commit()
    pass


def find_article(url_list: list[str] = None, session=None, only_saved=False, only_not_saved=False) -> list[Article]:
    if not session:
        session = Session()
    query = session.query(Article)
    if url_list:
        query = query.filter(Article.url.in_(url_list))
    if only_saved:
        query = query.filter(Article.saved_to_readwise == True)
    if only_not_saved:
        query = query.filter(Article.saved_to_readwise == False)

    return query.all()


if __name__ == '__main__':
    # mark_url_as_saved("https://www.google.com", agent="google", resp="{}")
    print(filter_saved_urls(["https://www.google.com", "https://www.google.com/1"]))

File: persistence.go
package main

import (
	"fmt"
	"gorm.io/driver/sqlite"
	"gorm.io/gorm"
	"gorm.io/gorm/schema"
	"os"
	"path/filepath"
	"strings"
	"time"
)

type Article struct {
	URL             string `gorm:"type:varchar(2048);primaryKey"`
	Agent           string `gorm:"type:varchar(128)"`
	SavedToReadwise bool   `gorm:"type:boolean"`
	SaveTime        string `gorm:"type:varchar(128)"`
	ReadwiseResp    string `gorm:"type:varchar(1024)"`
	// Content         string `gorm:"type:varchar"`
	ActualURL   string
	PublishTime time.Time `gorm:"type:datetime"`
	CreateTime  time.Time `gorm:"type:datetime;autoCreateTime"`
	UpdateTime  time.Time `gorm:"type:datetime;autoUpdateTime"`
}

var db *gorm.DB

func initDB() {
	var err error
	db, err = gorm.Open(sqlite.Open("data/readform.db"), &gorm.Config{NamingStrategy: schema.NamingStrategy{
		SingularTable: true,
	}})
	if err != nil {
		panic("failed to connect database")
	}
	err = db.AutoMigrate(&Article{})
	if err != nil {
		panic(err)
	}
}

func addArticle(url string, agent string, actualURL string) error {
	var articles []*Article
	err := db.Find(&articles, "url = ?", url).Error
	if err != nil {
		return err
	}
	if len(articles) == 0 {
		// Article does not exist, create a new one
		article := Article{
			URL:       url,
			ActualURL: actualURL,
			Agent:     agent,
		}
		return db.Create(&article).Error
	} else {
		// Article exists, update it
		return db.Model(&articles[0]).Updates(Article{Agent: agent}).Error
	}
}

func markURLAsSaved(url string, agent string, resp string) error {
	var article Article
	if err := db.First(&article, "url = ?", url).Error; err != nil {
		// Article does not exist, create a new one
		article = Article{
			URL:             url,
			Agent:           agent,
			SavedToReadwise: true,
			SaveTime:        time.Now().Format("2006-01-02 15:04:05"),
			ReadwiseResp:    resp,
		}
		return db.Create(&article).Error
	} else {
		// Article exists, update it
		return db.Model(&article).Updates(Article{
			SavedToReadwise: true,
			SaveTime:        time.Now().Format("2006-01-02 15:04:05"),
			Agent:           agent,
			ReadwiseResp:    resp,
		}).Error
	}
}

// findArticle finds article from database. Legacy versions of Readform does not have ActualURL field,
// so hasActualURL=true can filter out items created by legacy version.
func findArticle(urlList []string, onlySaved bool, onlyNotSaved bool, hasActualURL bool) ([]Article, error) {
	var articles []Article
	tx := db
	if urlList != nil {
		tx = tx.Where("url IN (?)", urlList)
	}
	if onlySaved {
		tx = tx.Where("saved_to_readwise = ?", true)
	}
	if onlyNotSaved {
		tx = tx.Where("saved_to_readwise = ?", false)
	}
	if hasActualURL {
		tx = tx.Where("actual_url != ''")
	}

	if err := tx.Find(&articles).Error; err != nil {
		return nil, err
	}

	return articles, nil
}

// filterOldURLs filter out saved URLs, returning unsaved URLs.
func filterOldURLs(urls []string) ([]string, error) {
	articles, err := findArticle(urls, true, false, false)
	if err != nil {
		return nil, fmt.Errorf("findArticle failed: %w", err)
	}
	existURLs := make(map[string]struct{}, len(articles))
	for _, a := range articles {
		existURLs[a.URL] = struct{}{}
	}

	var unsavedURLs []string
	for _, url := range urls {
		if _, exist := existURLs[url]; !exist {
			unsavedURLs = append(unsavedURLs, url)
		}
	}
	unsavedURLs = UniqStringSlice(unsavedURLs)
	return unsavedURLs, nil
}

func urlToLocalFilePath(url string) string {
	fileName := strings.ReplaceAll(url, "/", "_")
	fileName = strings.ReplaceAll(fileName, ":", "")
	filePath := "data/html/" + fileName + ".html"
	return filePath
}

// saveHTMLToLocalFile saves URL content to local file.
func saveHTMLToLocalFile(url, htmlContent string) error {
	filePath := urlToLocalFilePath(url)

	dir := filepath.Dir(filePath)
	if err := os.MkdirAll(dir, 0755); err != nil {
		return fmt.Errorf("creating directory failed: %w", err)
	}

	err := os.WriteFile(filePath, []byte(htmlContent), 0644)
	if err != nil {
		return fmt.Errorf("WriteFile failed: %w", err)
	}
	return nil
}

// readLocalHTMLFile gets URL content from local file.
func readLocalHTMLFile(url string) (string, error) {
	filePath := urlToLocalFilePath(url)
	data, err := os.ReadFile(filePath)
	if err != nil {
		return "", err
	}
	return string(data), nil
}
